40-car-detector/main.py

```diff
@@ -0,0 +1,227 @@
+    # 1970-01-01 00:00:00 UTC. So, the epoch used by the Notecard is ahead of
+    # the epoch used by MicroPython by 946684800 seconds. Nothing needs to be
+    # done here, unless one wants to print the time.
+    adjusted_time = notecard_time
+    gmtime = time.gmtime(adjusted_time)
+    year = gmtime[0]
+    month = gmtime[1]
+    day = gmtime[2]
+    hour = gmtime[3]
+    minute = gmtime[4]
+    second = gmtime[5]
+    weekday = gmtime[6]
+    rtc = RTC()
+    rtc.datetime((year, month, day, weekday, hour, minute, second, 0))
+
+    print(f'RTC time set.')
+
+    return rtc
+
+
+def main():
+    # Uncomment this line and replace com.your-company:your-product-name with your
+    # ProductUID.
+    # product_uid = 'com.your-company:your-product-name'
+    card = setup_notecard(product_uid)
+    rtc = setup_rtc(card)
+
+    env_var_names = ['publish_period', 'confidence_threshold']
+    config = refresh_config(fetch_env_vars(card, env_var_names))
+    last_env_refresh_ms = time.ticks_ms()
+    env_refresh_period_ms = 3 * 60 * 1000
+
+    setup_camera()
+
+    labels, net = tf.load_builtin_model('trained')
+    clock = time.clock()
+
+    event_start_threshold = 1
+    event_end_threshold = 12
+    consecutive_detections = 0
+    consecutive_no_detections = 0
+    event_count = 0
+    event_happening = False
+
+    last_publish_ms = time.ticks_ms()
+    while True:
+        img = sensor.snapshot()
+
+        for i, detection_list in enumerate(
+                net.detect(img,
+                           thresholds=[
+                               (math.ceil(config.confidence_threshold * 255),
+                                255)
+                           ])):
+            if (i == 0): continue  # background class
+            if (len(detection_list) == 0):
+                consecutive_detections = 0
+                consecutive_no_detections += 1
+            else:
+                consecutive_detections += 1
+                consecutive_no_detections = 0
+
+                for d in detection_list:
+                    [x, y, w, h] = d.rect()
+                    center_x = math.floor(x + (w / 2))
+                    center_y = math.floor(y + (h / 2))
+                    print('Car detected @ (%d, %d).' % (center_x, center_y))
+                    img.draw_circle((center_x, center_y, 12), thickness=2)
+
+        if consecutive_detections >= event_start_threshold and not event_happening:
+            event_start_time = rtc.datetime()
+            print('Event started.')
+            event_happening = True
+            event_count += 1
+
+        if event_happening and consecutive_no_detections >= event_end_threshold:
+            event_end_time = rtc.datetime()
+            print(f'Event ended. Event count is now {event_count}.')
+            event_happening = False
+            add_event(card, event_start_time, event_end_time)
+
+        if time.ticks_diff(time.ticks_ms(),
```

Comment:
I'm curious why we do this in code compared to say using `hub.set:outbound`.

Reply:
Reworking to use `hub.set`.